libnetwork/resolver.go

```diff
@@ -393,199 +363,200 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
+		logrus.WithError(err).Errorf("[resolver] failed to handle query: %s (%s) from %s", queryName, dns.TypeToString[queryType], w.RemoteAddr().String())
+		reply(new(dns.Msg).SetRcode(query, dns.RcodeServerFailure))
 		return
 	}
 
-	if resp == nil {
-		// If the backend doesn't support proxying dns request
-		// fail the response
-		if !r.proxyDNS {
-			resp = new(dns.Msg)
-			resp.SetRcode(query, dns.RcodeServerFailure)
-			if err := w.WriteMsg(resp); err != nil {
-				logrus.WithError(err).Error("[resolver] error writing dns response")
+	if resp != nil {
+		// We are the authoritative DNS server for this request so it's
+		// on us to truncate the response message to the size limit
+		// negotiated by the client.
+		maxSize := dns.MinMsgSize
+		if w.LocalAddr().Network() == "tcp" {
+			maxSize = dns.MaxMsgSize
+		} else {
+			if optRR := query.IsEdns0(); optRR != nil {
+				if udpsize := int(optRR.UDPSize()); udpsize > maxSize {
+					maxSize = udpsize
+				}
 			}
-			return
 		}
+		resp.Truncate(maxSize)
+		reply(resp)
+		return
+	}
 
+	if r.proxyDNS {
 		// If the user sets ndots > 0 explicitly and the query is
 		// in the root domain don't forward it out. We will return
 		// failure and let the client retry with the search domain
-		// attached
-		switch queryType {
-		case dns.TypeA, dns.TypeAAAA:
-			if r.backend.NdotsSet() && !strings.Contains(strings.TrimSuffix(queryName, "."), ".") {
-				resp = createRespMsg(query)
-			}
+		// attached.
+		if (queryType == dns.TypeA || queryType == dns.TypeAAAA) && r.backend.NdotsSet() &&
+			!strings.Contains(strings.TrimSuffix(queryName, "."), ".") {
+			resp = createRespMsg(query)
+		} else {
+			resp = r.forwardExtDNS(w.LocalAddr().Network(), query)
 		}
 	}
 
-	proto := w.LocalAddr().Network()
-	maxSize := 0
-	if proto == "tcp" {
-		maxSize = dns.MaxMsgSize - 1
-	} else if proto == "udp" {
-		optRR := query.IsEdns0()
-		if optRR != nil {
-			maxSize = int(optRR.UDPSize())
-		}
-		if maxSize < defaultRespSize {
-			maxSize = defaultRespSize
-		}
+	if resp == nil {
+		// We were unable to get an answer from any of the upstream DNS
+		// servers or the backend doesn't support proxying DNS requests.
+		resp = new(dns.Msg).SetRcode(query, dns.RcodeServerFailure)
 	}
+	reply(resp)
+}
 
-	if resp != nil {
-		if resp.Len() > maxSize {
-			truncateResp(resp, maxSize, proto == "tcp")
+func (r *Resolver) dialExtDNS(proto string, server extDNSEntry) (net.Conn, error) {
+	var (
+		extConn net.Conn
+		dialErr error
+	)
+	extConnect := func() {
+		if server.port == 0 {
+			server.port = 53
 		}
-	} else {
-		for i := 0; i < maxExtDNS; i++ {
-			extDNS := &r.extDNSList[i]
-			if extDNS.IPStr == "" {
-				break
-			}
-			extConnect := func() {
-				addr := fmt.Sprintf("%s:%d", extDNS.IPStr, 53)
-				extConn, err = net.DialTimeout(proto, addr, extIOTimeout)
-			}
+		addr := fmt.Sprintf("%s:%d", server.IPStr, server.port)
```

Comment:
This'll do the wrong thing with IPv6 addresses. Use `net.JoinHostPort`

Reply:
On second thought, I would rather leave this change for another PR dedicated to adding support for external IPv6 DNS servers. Every code path which sets the external servers filters out IPv6 addresses.
https://github.com/moby/moby/blob/4e276400760585c9d5969488c9b44e5176a58c18/libnetwork/sandbox_dns_unix.go#L261
https://github.com/moby/moby/blob/4e276400760585c9d5969488c9b44e5176a58c18/libnetwork/sandbox_dns_unix.go#L267
https://github.com/moby/moby/blob/4e276400760585c9d5969488c9b44e5176a58c18/libnetwork/sandbox_dns_unix.go#L402